File: Code-spark/services/common-library/src/main/java/com/dao/common/notification/NotificationProducerService.java
package com.dao.common.notification;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;

import java.time.Instant;

@Service
public class NotificationProducerService {

    private static final Logger LOGGER = LoggerFactory.getLogger(NotificationProducerService.class);
    private final KafkaTemplate<String, String> kafkaTemplate;
    private final ObjectMapper objectMapper;
    private static final String TOPIC = "notifications";

    // Khởi tạo (Constructor)
    public NotificationProducerService(KafkaTemplate<String, String> kafkaTemplate, ObjectMapper objectMapper) {
        this.kafkaTemplate = kafkaTemplate;
        this.objectMapper = objectMapper;
    }

    // Hàm gửi thông báo
    public void sendNotification(NotificationMessage message) {
        if (message.getCreatedAt() == null) {
            message.setCreatedAt(Instant.now());
        }
        try {
            String payload = objectMapper.writeValueAsString(message);
            kafkaTemplate.send(TOPIC, payload);
            LOGGER.info("Gửi thông báo thành công cho user: {}", message.getRecipientUserId());
        } catch (JsonProcessingException e) {
            LOGGER.error("Lỗi khi chuyển đổi JSON: {}", e.getMessage());
        }
    }
}

File: Code-spark/services/common-node/src/notification/NotificationProducerService.js
const { Kafka } = require('kafkajs');

class NotificationProducerService {
  constructor(brokers, clientId = 'node-service') {
    this.kafka = new Kafka({ clientId, brokers });
    this.producer = this.kafka.producer();
    this.topic = 'notifications';
  }

  async connect() {
    await this.producer.connect();
    console.log(`[Kafka] Đã kết nối Producer: ${this.kafka.clientId}`);
  }

  async sendNotification(messageObj) {
    if (!messageObj.createdAt) {
      messageObj.createdAt = new Date().toISOString();
    }

    try {
      const payload = JSON.stringify(messageObj);
      await this.producer.send({
        topic: this.topic,
        messages: [{ value: payload }]
      });
      console.log(`[Kafka] Gửi thông báo thành công cho user: ${messageObj.recipientUserId}`);
    } catch (error) {
      console.error("[Kafka] Lỗi khi gửi tin nhắn:", error.message);
    }
  }

  async disconnect() {
    await this.producer.disconnect();
  }
}
module.exports = NotificationProducerService;